File: lib/realtime/data/vehicle_position.ex
defmodule Realtime.Data.VehiclePosition do
  @moduledoc """
  The subset of the VehiclePosition data in the enhanced_json feed that we use.
  """
  use Ecto.Schema

  @type t :: %__MODULE__{
          route_id: Realtime.Data.route_id(),
          revenue: boolean(),
          direction: integer(),
          label: String.t(),
          cars: [String.t()],
          position: Util.Position.t(),
          heading: float() | nil,
          station_id: String.t() | nil,
          stop_id: String.t() | nil,
          current_status: :INCOMING_AT | :STOPPED_AT | :IN_TRANSIT_TO,
          timestamp: DateTime.t(),
          # GTFS-RT id, e.g. "R-54826B16".
          vehicle_id: String.t() | nil,
          trip_id: String.t() | nil
        }

  @derive {Jason.Encoder,
           only: [
             :route_id,
             :revenue,
             :direction,
             :label,
             :cars,
             :position,
             :heading,
             :station_id,
             :stop_id,
             :current_status,
             :timestamp,
             :vehicle_id,
             :trip_id
           ]}
  schema "vehicle_positions" do
    field(:route_id, Util.AtomType)
    field(:revenue, :boolean, default: true)
    field(:direction, :integer)
    field(:label, :string)
    field(:cars, {:array, :string})
    field(:position, Util.PositionType)
    field(:heading, :float, default: nil)
    field(:station_id, :string, default: nil)
    field(:stop_id, :string, default: nil)
    field(:current_status, Util.AtomType, default: :IN_TRANSIT_TO)
    field(:timestamp, :utc_datetime)
    field(:vehicle_id, :string, virtual: true, default: nil)
    field(:trip_id, :string, virtual: true, default: nil)
    timestamps(type: :utc_datetime)
  end
end

File: js/models/vehiclePosition.ts
import { dateTimeFromISO } from "../dateTime";
import { CarId, DirectionId, RouteId, StationId } from "./common";
import { LatLng } from "./latlng";
import { DateTime } from "luxon";
import { z } from "zod";

export enum StopStatus {
  InTransitTo,
  StoppedAt,
}

const StopStatusData = z.enum(["INCOMING_AT", "STOPPED_AT", "IN_TRANSIT_TO"]);
type StopStatusData = z.infer<typeof StopStatusData>;

export const stopStatusFromData = (data: StopStatusData): StopStatus => {
  switch (data) {
    case "INCOMING_AT":
    case "IN_TRANSIT_TO":
      return StopStatus.InTransitTo;
    case "STOPPED_AT":
      return StopStatus.StoppedAt;
  }
};

export const VehiclePositionData = z.object({
  route_id: z.string(),
  revenue: z.boolean(),
  direction: z.number(),
  label: z.string(),
  cars: z.array(z.string()),
  position: z
    .object({ latitude: z.number(), longitude: z.number() })
    .nullable(),
  heading: z.number().nullable(),
  station_id: z.string().nullable(),
  stop_id: z.string().nullable(),
  current_status: StopStatusData,
  timestamp: z.string().nullable(),
  vehicle_id: z.string().nullable(),
  trip_id: z.string().nullable(),
});
export type VehiclePositionData = z.infer<typeof VehiclePositionData>;

export type VehiclePosition = {
  routeId: RouteId;
  revenue: boolean;
  directionId: DirectionId;
  label: CarId;
  cars: CarId[];
  position: LatLng | null;
  stationId: StationId | null;
  stopId: string | null;
  stopStatus: StopStatus;
  heading: number | null;
  timestamp: DateTime | null;
  vehicleId: string | null;
  tripId: string | null;
};

export const vehiclePositionFromData = (
  data: VehiclePositionData,
): VehiclePosition => ({
  routeId: data.route_id,
  revenue: data.revenue,
  directionId: data.direction,
  label: data.label,
  cars: data.cars,
  stationId: data.station_id,
  stopId: data.stop_id,
  stopStatus: stopStatusFromData(data.current_status),
  position: data.position,
  heading: data.heading,
  timestamp: data.timestamp !== null ? dateTimeFromISO(data.timestamp) : null,
  vehicleId: data.vehicle_id,
  tripId: data.trip_id,
});
